budgetkey_data_pipelines/pipelines/budget/national/changes/processed/extract-change-groups.py

```diff
@@ -0,0 +1,288 @@
+
+
+def combinations(iterable, r):
+    # combinations('ABCD', 2) --> AB AC AD BC BD CD
+    # combinations(range(4), 3) --> 012 013 023 123
+    def get_next(next_of, i):
+        ret = nexts.get(next_of, next_of + 1)
+        if ret > i + n - r:
+            ret = i + n - r
+        return ret
+
+    pool = tuple(iterable)
+    n = len(pool)
+    if r > n:
+        return
+    nexts = dict(zip(range(n - 1), range(1, n)))
+    indices = list(range(r))
+    while True:
+        found = (yield tuple(pool[i] for i in indices))
+        if found:
+            for i in indices:
+                target = nexts.get(i)
+                sources = (k for k, v in nexts.items() if v == i)
+                if target is not None:
+                    for src in sources:
+                        nexts[src] = target
+                else:
+                    for src in list(sources):
+                        del nexts[src]
+            indices[0] = get_next(indices[0], 0)
+            for j in range(1, r):
+                indices[j] = get_next(indices[j - 1], j)
+        for i in reversed(range(r)):
+            if indices[i] != i + n - r:
+                break
+        else:
+            return
+        indices[i] = get_next(indices[i], i)
+        for j in range(i + 1, r):
+            indices[j] = get_next(indices[j - 1], j)
+
+
+def get_changes(rows):
+    result = []
+
+    for row in rows:
+        change = dict(row)
+
+        # if change['year'] is None:
+        #     change['year'] = datetime.strptime(
+        #         change['date/approval'], '%d/%m/%Y').year
+
+        change['trcode'] = transfer_code(change)
+        change['_value'] = sum(change.get(x, 0) for x in value_fields)
+        for k, v in change.items():
+            if k.startswith('date'):
+                if v is None:
+                    v = datetime.now().strftime('%d/%m/%Y')
+                    kind = 'pending'
+                    change['date/pending'] = v
+                    change['pending'] = True
+                else:
+                    kind = k[5:]
+                change['date'] = v
+                change['pending'] = kind == 'pending'
+                change['date_kind'] = kind + '/' + v
+                break
+        result.append(change)
+
+    return result
+
+
+def get_groups(changes):
+    def get_date(x):
+        return x['date_kind']
+
+    def get_keys_for_prefix(ch, prefix):
+        return set(
+            (x['leading_item'], x['req_code'])
+            for x in ch if x['budget_code'].startswith(prefix)
+        )
+
+    def initialize_groups(changes):
+        result = []
+        selected_transfer_codes = set()
+        for date_kind, date_changes in itertools.groupby(changes, get_date):
+            date_changes = list(date_changes)
+            date = date_changes[0]['date']
+            date_reserve = [
+                c for c in date_changes
+                if c['budget_code'].startswith('0047') and c['leading_item'] != 47
```

Comment:
It is - but the data had a problem (started with '47' instead of '0047').
I fixed that now.